application_store/db/queries/application/queries.py

```diff
@@ -266,25 +269,96 @@ def submit_application(application_id) -> Applications:
             "jsonb_blob": ApplicationSchema().dump(application),
             "type_of_application": application_type,
         }
-        stmt = postgres_insert(AssessmentRecord).values([row])
 
-        upsert_rows_stmt = stmt.on_conflict_do_nothing(index_elements=[AssessmentRecord.application_id]).returning(
-            AssessmentRecord.application_id
+        existing_application = db.session.scalar(
+            select(AssessmentRecord)
+            .where(
+                AssessmentRecord.application_id == row["application_id"],
+                AssessmentRecord.is_withdrawn == False,  # noqa: E712
+            )
+            .options(load_only(AssessmentRecord.jsonb_blob))
         )
 
-        result = db.session.execute(upsert_rows_stmt)
-
-        # Check if the inserted application is in result
-        inserted_application_ids = [item.application_id for item in result]
-        if not len(inserted_application_ids):
-            current_app.logger.warning(
-                "Application already exists in the database: {app_id}", extra=dict(app_id=row["application_id"])
-            )
+        if existing_application and fund.funding_type == "UNCOMPETED":
+            # For uncompeted funds, the application may already exist and this may be a resubmission.
+            field_map = {}
+            changed_fields = set()
+            for form in existing_application.jsonb_blob["forms"]:
+                for section in form["questions"]:
+                    for field in section["fields"]:
+                        field_map[field["key"]] = field["answer"]
+
+            for form in row["jsonb_blob"]["forms"]:
+                for section in form["questions"]:
+                    for field in section["fields"]:
+                        if field["answer"] != field_map.get(field["key"]):
+                            changed_fields.add(field["key"])
+                            if "history_log" in field:
+                                field["history_log"].append(
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                )
+                            else:
+                                field["history_log"] = [
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                ]
+
+            if changed_fields:
+                row["workflow_status"] = WorkflowStatus.CHANGE_RECEIVED
+
+            stmt = postgres_insert(AssessmentRecord).values(row)
+
+            # setting with an update makes sure the derived values are recalculated
+            update_row_statement = stmt.on_conflict_do_update(
+                index_elements=[AssessmentRecord.application_id], set_=row
+            ).returning(AssessmentRecord.application_id)
+
+            db.session.execute(update_row_statement)
+
+            flags = db.session.scalars(
```

Comment:
Similar here. I think this relation is already defined, so it can be assessment_flag.flags.

If you want to apply filter on it, there are a number of ways to go about it:
1. You can fetch all and iff inside the for loop (bit nasty)
2. You can add a `hybrid_property` inside the model:
```
    @hybrid_property
    def unfinished_flags(self):
        return [flag for flag in self.flags if not flag.is_finished]
```
Still similar to 1. but the query is done inside the model.
3. (the most efficient one) define the flags relation as:
```
    flags = relationship(
        "AssessmentFlag",
        lazy="dynamic",  # Makes it queryable
    )
```
So you should be able to do something like this:
```
unfinished_flags = assessment_record.flags.filter_by(is_finished=False).all()
```

Reply:
I ended up adding a new relationship `change_requests` that filters on the join - I think that'll come in handy for the work we're doing.